kythe/java/com/google/devtools/kythe/analyzers/java/JavaIndexer.java

```diff
@@ -83,53 +89,74 @@ public static void main(String[] args) throws AnalysisException, IOException {
     if (!Strings.isNullOrEmpty(config.getIndexPackRoot())) {
-      // java_indexer --index_pack=archive-root unit-key
-      desc = new Archive(config.getIndexPackRoot()).readDescription(compilationPath);
+      // java_indexer --index_pack=archive-root unit-key+
+      logger.atWarning().log(
+          "index pack files are deprecated; "
+              + "use kzip files instead: https://kythe.io/docs/kythe-kzip.html");
+      Archive archive = new Archive(config.getIndexPackRoot());
+      for (String unitDigest : config.getCompilation()) {
+        CompilationDescription desc = archive.readDescription(unitDigest);
+        analyzeCompilation(config, plugins, statistics, desc);
+      }
     } else {
-      if (compilationPath.endsWith(IndexInfoUtils.KZIP_FILE_EXT)) {
-        // java_indexer kzip-file
-        try {
-          Collection<CompilationDescription> compilationDescriptions =
-              IndexInfoUtils.readKZip(compilationPath);
-          if (config.getIgnoreEmptyKZip() && compilationDescriptions.isEmpty()) {
-            return;
+      // java_indexer compilation-file+
+      boolean loggedDeprecationWarning = false;
+      for (String compilationPath : config.getCompilation()) {
+        if (compilationPath.endsWith(IndexInfoUtils.KZIP_FILE_EXT)) {
+          // java_indexer kzip-file
+          boolean foundCompilation = false;
+          try {
+            KZip.Reader reader = new KZipReader(new File(compilationPath));
+            for (IndexedCompilation indexedCompilation : reader.scan()) {
+              foundCompilation = true;
+              CompilationDescription desc =
+                  IndexInfoUtils.indexedCompilationToCompilationDescription(
+                      indexedCompilation, reader);
+              analyzeCompilation(config, plugins, statistics, desc);
+            }
+          } catch (KZipException e) {
+            throw new IllegalArgumentException("Unable to read kzip", e);
           }
-          if (compilationDescriptions.size() != 1) {
+          if (!config.getIgnoreEmptyKZip() && !foundCompilation) {
             throw new IllegalArgumentException(
-                "The kzip did not contain exactly 1 CompilationDescription. It contained "
-                    + compilationDescriptions.size());
+                "given empty .kzip file \"" + compilationPath + "\"; try --ignore_empty_kzip");
           }
-          desc = compilationDescriptions.iterator().next();
-        } catch (KZipException e) {
-          throw new IllegalArgumentException("Unable to read kzip", e);
-        }
-      } else {
-        // java_indexer kindex-file
-        try {
-          desc = IndexInfoUtils.readKindexInfoFromFile(compilationPath);
-        } catch (EOFException e) {
-          if (config.getIgnoreEmptyKIndex()) {
-            return;
+        } else {
+          // java_indexer kindex-file
+          if (!loggedDeprecationWarning) {
+            logger.atWarning().log(
```

Comment:
`logger.atWarning().atMostEvery(1, DAYS)` ? I wish there was a clearer way to say `logger.atWarning().once()`, but given the expected duration of a single extraction, using DAYS seems fine and will avoid an extraneous logging local.  But up to you if you think it's worth it.

Reply:
Particularly since you're never actually setting the local bool ;-)

Reply:
There is a Google-internal method that would work for `once()`, but...